core/store/src/flat_state.rs

```diff
@@ -726,6 +870,92 @@ mod tests {
         assert_eq!(delta.get(&[5]), Some(Some(ValueRef::new(&[9]))));
     }
 
+    // This test tests some basic use cases for FlatState and FlatStorageState.
+    // We created a linear chain with no forks, start with flat head at the genesis block, then
+    // moves the flat head forward, which checking that flat_state.get_ref() still returns the correct
+    // values and the state is being updated in store.
+    #[test]
+    fn flat_storage_state_sanity() {
+        // 1. Create a chain with 10 blocks with no forks. Set flat head to be at block 0.
+        //    Block i sets value for key &[1] to &[i].
+        let mut chain = MockChain::linear_chain(10);
+        let store = create_test_store();
+        let mut store_update = store.store_update();
+        store_helper::set_flat_head(&mut store_update, 0, &chain.get_block_hash(0));
```

Comment:
Can we check in `set_flat_head` that flat state head is not set previously, or add a comment that it must be called only once? Though I am not sure how it will work with catchup logic.

Adding comment here because `set_flat_head` is unchanged in the PR.

Reply:
I'm not sure I understand what you are referring to. `set_flat_head` can be called many times, it is called every time when flat_head is updated. Are you worried that the function `set_flat_head` can be arbitrarily called in the code, not through `update_flat_head`?

Reply:
> Are you worried that the function set_flat_head can be arbitrarily called in the code, not through update_flat_head?

Yeah, that's right. I'm thinking about making `set_flat_head` private, or explicitly saying that it must be called only inside `update_flat_head` or on initialization.

Reply:
I made it private and added another function `set_flat_storage_state_for_genesis`